### PR TITLE
Additional EH tests

### PR DESCRIPTION
**Pull request by russellhadley**
Adding some more selected EH tests.  At this point I'm still not including
particular tests that were repros for MSVC compiler bugs since more work is
needed to tell if these are useful.

### CODE REVIEW

**Comment by russellhadley** [comment]
@majnemer - PTAL

File: eh/ehthrow.cxx [hunk 1]
```diff
@@ -0,0 +1,547 @@
...
+        throw b;
+    }catch(...){
+        puts("In catch #1");
+        foo10();
+    }
+    puts("End of test10, throwing a A");
+    throw A();
+}
+
+void main()
```
Comment (majnemer) [inline eh/ehthrow.cxx]: This will not compile with clang, can you make the return type `int`?

File: eh/nesttry.cxx [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    ~A(){printf("In ~A()\n");}
+};
+
+void foo()
+{
+    printf("Throwing\n");
+    throw 1;
+}
+
+void main()
```
Comment (majnemer) [inline eh/nesttry.cxx]: Ditto here.

File: eh/noreturn.cpp [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    throw 1;
+}
+
+void foo(A& a, int i)
+{
+    A a1 = a;
+    bar(a1);
+}
+
+void main()
```
Comment (majnemer) [inline eh/noreturn.cpp]: Ditto here.

File: eh/recursive_throw.cpp [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        }
+    }
+    else
+    {
+        printf("Level MAX_REC throwing.\n");
+        throw level;
+    }
+}
+
+void main(void)
```
Comment (majnemer) [inline eh/recursive_throw.cpp]: Ditto here.

File: eh/noreturn.cpp [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    throw 1;
+}
+
+__declspec(noreturn)
+A foobar()
+{
+    throw 1;
+}
+
+void foo(A& a, int i)
```
Comment (majnemer) [inline eh/noreturn.cpp]: The `a` parameter needs to be a `const` l-value reference: `const A& a`.

File: eh/terminate.cpp [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+{
+public:
+    C() {}
+    ~C()
+    {
+        printf( "in C::~C()\n");
+        if (FGenSEHException)
+        {
+            printf("generating access violation (SEH)\n");
+            *(char*)(0) = 0; // Uh, EHa, don't you think?
```
Comment (majnemer) [inline eh/terminate.cpp:33]: This should be a volatile store: `*(volatile char*)(0) = 0;`

**Comment by majnemer** [comment]
Looks good, found some trivial issues.

**Comment by russellhadley** [comment]
@majnemer made the changes you suggested.  Merging now.

**Pull request merged**